### ISSUE TITLE
bwasw extremely slow on plant genomes

### ISSUE THREAD

**Issue by maximilianh** [lh3/bwa#230]
Even for a 100bp input sequence, with repeat-rich plant genomes, bwasw will not complete. I can provide an example test case but if you just run bwasw onto any plant genome you should see the problem. I'd love to use an option to reduce the search depth, but the various options to optimize the seed search (?) are not documented anywhere.

**Comment by andreasbuhr** [comment]
Hi, I am currently profiling some parts of bwa, but I am not familiar with bioinformatics.
Could you provide a step-by-step manual to reproduce this behaviour? So what files to download where and what command to run on them?
Thank you!

**Comment by maximilianh** [comment]
I am taking random big plant genome, e.g. GCA_900519105.1 from NCBI. 

wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/900/519/105/GCA_900519105.1_iwgsc_refseqv1.0/GCA_900519105.1_iwgsc_refseqv1.0_genomic.fna.gz
gunzip GCA_900519105.1_iwgsc_refseqv1.0_genomic.fna.gz
bwa index GCA_900519105.1_iwgsc_refseqv1.0_genomic.fna

cat <<' EOF' > test.fa
>test
AGTTACCACAGGGATAACTGGCTTGTGGCAGCCAAGCGTTCATAGCGACG
TTGCTTTTTGATCCTTTGATGTCGGCTCTTCCTATCATTGTGAAGCAGAA
TTCACCAAGTGTTGGATTGTTCACCCACCAATAGGGAACGTGAGCTGGGT
TTAGACCGTCGTGAGACAGGTTAGTTTTACCCTACTGATGACAGTGTCGC
GATAGTAATTCAACCTAGTACAAGAGGAACCATTGATTCACACAATTGGT
CATCGCGCTTGGTTGAAAAGCCAGTGGCGTGAAGCTACCGTGTGCCGGAT
TATGACTGAACGCCTCTAAGTCAGAATCCAAGCTAGCATGCGACGCCTGC
GCCCGCCGCCCGCCCTGAGCCACGTTAGGGGCGCTTGCGCCCCCAAGGGC
CCGTGCCATTGGCTAAGCCGGTCCGGCCGACGTGCCGCGGCCGGCCGCCT
EOF

bwa bwasw GCA_900519105.1_iwgsc_refseqv1.0_genomic.fna test.fa

I killed this after five minutes without any output on my machine with 64GB of RAM. 

Any idea on how I could speed this up (index in a different way? any secret arguments to make this faster? I've tried lowering the gapsize to 0 and the bandwith (what is "bandwidth"?) and tried all other arguments. 

On the UCSC email list, we got someone who has benchmarked various aligners and says that the fastest one for plants is GSAlign. Will look into that.

**Comment by maximilianh** [comment]
I'm curious... if you're not familiar with this software... why are you profiling it...?

**Comment by lh3** [comment]
bwasw is largely deprecated by bwa-mem.

**Comment by andreasbuhr** [comment]
Thanks a lot for your reply!

> I'm curious... if you're not familiar with this software... why are you profiling it...?

I am caring about software performance. bwa makes an excellent example for software profiling training classes, as the codebase is rather small, it compiles quickly, there are a lot of performance-related low-level tricks to be seen in the code, and it is a performance critical application.

**Comment by maximilianh** [comment]
@lh3 Thanks for your reply. I know, but I need an aligner for long ~2kbp input sequences. Do you still think I should use bwa-mem for that?

**Comment by lh3** [comment]
use minimap2

**Comment by maximilianh** [comment]
@andreasbuhr That sounds great. In this case, you may want to rather improve bwa-mem or minimap, as these are the software packages that most people use. My use case is relatively unusual, I first align a 2kbp input piece to the genome to find the 100%-identical match, then I use bwa-aln to find all matches for every 20bp part in it.

**Comment by maximilianh** [comment]
@lh3 The problem is that users input sequences of length 20bp to 2000bp on my website crispor.org (I described my problem around four years ago in an email to you, you didn't reply, which I understand, but I could send the email again). So I need an aligner that can map an input sequence to the genome and usually finds the best match, but some sequences can be relatively short. Would minimap be good for that? 

One alternative is to use a combination: bwa-mem for sequences <250bp and minimap for longer ones?

**Comment by lh3** [comment]
Yes, you should use a combination. Bwa-aln is still better for <32bp.

**Comment by maximilianh** [comment]
Yes, but using bwa-aln to find 27bp in this 14GB genome takes quite a bit
of time, too (64GB RAM, 8 CPUs, local hard disk):



[main] CMD: bwa aln GCA_900519105.1.fa /tmp/temp.fa

[main] Real time: 122.985 sec; CPU: 19.723 sec


real    2m2.986s

user    0m0.000s

sys     0m19.723s

>
